hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/CryptoApproveAllowanceTransactionHandler.java

```diff
@@ -84,11 +96,27 @@ public void updateTransaction(Transaction transaction, RecordItem recordItem) {
             cryptoAllowance.setOwner(ownerAccountId.getId());
             cryptoAllowance.setPayerAccountId(payerAccountId);
             cryptoAllowance.setSpender(EntityId.of(cryptoApproval.getSpender()).getId());
-            cryptoAllowance.setTimestampLower(consensusTimestamp);
-            entityListener.onCryptoAllowance(cryptoAllowance);
+            cryptoAllowance.setTimestampLower(recordItem.getConsensusTimestamp());
+
+            if (!cryptoAllowanceState.containsKey(cryptoAllowance.getId())) {
+                entityListener.onCryptoAllowance(cryptoAllowance);
+                cryptoAllowanceState.put(cryptoAllowance.getId(), cryptoAllowance);
+            }
```

Comment:
Could optimize to only search the map once.

```suggestion
            if (cryptoAllowanceState.putIfAbsent(cryptoAllowance.getId(), cryptoAllowance) == null) {
                entityListener.onCryptoAllowance(cryptoAllowance);
            }
```